public/src/card/sectioned-card-list.html

```diff
@@ -0,0 +1,211 @@
+            observer: '_refreshMore'
+          },
+          label: String
+        };
+      }
+
+      constructor() {
+        super();
+        this._resizeListener = _debounce(() => {
+          this._onResize();
+        }, 350);
+      }
+
+      disconnectedCallback() {
+        super.disconnectedCallback();
+        window.removeEventListener("resize", this._resizeListener);
+      }
+
+      _refreshMore() {
+        if (this.more) {
+          this.$.more.classList.remove("hidden");
+        } else {
+          this.$.more.classList.add("hidden");
+        }
+      }
+
+      _onMore(event) {
+        event.stopPropagation();
+        this.dispatchEvent(new CustomEvent('more', { bubbles: true, composed: true }));
+      }
+
+      getPanelClass(item) {
+        return item.promoted ? "promotedPanel" : "";
+      }
+
+      _onResize() {
+        if (this.sections && this.sections.length) {
+          let colsPerRow = Math.min(4, Math.floor(this.offsetWidth / 312));
+          if (colsPerRow != this._colsPerRow) {
+            this._relayout();
+          }
+        }
+      }
+
+      _relayout() {
+        this.sections = this._createSections(this.items, true);
+      }
+
+      _createSections(items, useLabel) {
+        let colsPerRow = Math.min(4, Math.floor(this.offsetWidth / 312));
+        this._colsPerRow = colsPerRow;
+        let list = [];
+        let plist = [];
+        let current = { cards: [], more: false, label: useLabel ? this.label : "" };
+        let pcurrent = { label: 'Promoted', cards: [], more: false, promoted: true };
+        for (let i = 0; i < items.length; i++) {
+          let item = items[i];
+          if (item.promoted) {
+            if (pcurrent.cards.length === 0) {
+              plist.push(pcurrent);
+            } else if (pcurrent.cards.length === colsPerRow) {
+              pcurrent = { label: 'Promoted', cards: [], more: false, promoted: true };
+              plist.push(pcurrent);
+            }
+            pcurrent.cards.push(item);
+          } else {
+            if (current.cards.length === 0) {
+              list.push(current);
+            } else if (current.cards.length === colsPerRow) {
+              current = { cards: [], more: false };
+              list.push(current);
+            }
+            current.cards.push(item);
+          }
+        }
+        let sections = [];
+        while (true) {
+          let a = list.shift();
+          let b = plist.shift();
+          if ((!a) && (!b)) {
+            break;
+          }
+          if (a) {
+            sections.push(a);
+          }
+          if (b) {
+            sections.push(b);
+          }
+        }
+        return sections;
+      }
+
+      _onItems() {
+        window.removeEventListener("resize", this._resizeListener);
+        if (this.items && this.items.length) {
+          this._relayout();
+          window.addEventListener("resize", this._resizeListener);
+        } else {
+          this.sections = [];
```

Comment:
If items get appended later (using append()), isn't the resize listener going to be missing?

Reply:
Append calls "set" when there no existing items - which will trigger this